File: scripts/import_workflows.py
#!/usr/bin/env python3
import os
import json
import argparse
import urllib.request
import urllib.error
import sys

def find_workflow_files(root_dir):
    workflow_files = []
    for root, dirs, files in os.walk(root_dir):
        for file in files:
            if file == 'workflow.json' or file == 'orchestrator.json' or file == 'manager.json':
                workflow_files.append(os.path.join(root, file))
            # Also catch specific named orchestrators if they don't follow the exact naming convention but are json
            elif file.endswith('.json') and ('orchestrator' in file or 'workflow' in file):
                 if os.path.join(root, file) not in workflow_files: # avoid duplicates
                    workflow_files.append(os.path.join(root, file))
    return workflow_files

def get_existing_workflows(base_url, api_key):
    url = f"{base_url.rstrip('/')}/api/v1/workflows"
    headers = {
        'X-N8N-API-KEY': api_key,
        'Content-Type': 'application/json'
    }
    try:
        req = urllib.request.Request(url, headers=headers)
        with urllib.request.urlopen(req) as response:
            if response.status == 200:
                data = json.loads(response.read().decode('utf-8'))
                return {w['name'] for w in data.get('data', [])}
    except Exception as e:
        print(f"⚠️  Could not fetch existing workflows: {e}")
    return set()


def import_workflow(file_path, base_url, api_key, dry_run=False):
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            workflow_data = json.load(f)
    except Exception as e:
        print(f"❌ Error reading {file_path}: {e}")
        return False

    # Basic validation: check if it looks like an n8n workflow
    if 'nodes' not in workflow_data:
        print(f"⚠️  Skipping {file_path}: Does not look like a valid n8n workflow (missing nodes).")
        return False

    workflow_name = workflow_data.get('name', os.path.basename(file_path))
    print(f"📦 Found workflow: {workflow_name} ({file_path})")

    if dry_run:
        return True

    url = f"{base_url.rstrip('/')}/api/v1/workflows"
    headers = {
        'X-N8N-API-KEY': api_key,
        'Content-Type': 'application/json'
    }

    # Prepare payload for API
    # 1. Remove 'id' creating a new workflow usually generates one.
    # 2. Ensure 'settings' exists.
    # 3. Keep only standard fields to avoid "additional properties" error.
    
    valid_keys = ['name', 'nodes', 'connections', 'settings', 'meta']
    payload = {}
    
    for key in valid_keys:
        if key in workflow_data:
            payload[key] = workflow_data[key]
            
    # Ensure settings exists
    if 'settings' not in payload:
        payload['settings'] = {}
        
    # If original had 'name', ensure it is used, otherwise filename
    if 'name' not in payload:
        payload['name'] = workflow_name

    try:
        req = urllib.request.Request(url, data=json.dumps(payload).encode('utf-8'), headers=headers, method='POST')
        with urllib.request.urlopen(req) as response:
            if response.status in [200, 201]:
                res_json = json.loads(response.read().decode('utf-8'))
                new_id = res_json.get('id')
                print(f"✅ Successfully imported '{workflow_name}' -> ID: {new_id}")
                return True
            else:
                print(f"❌ Failed to import '{workflow_name}'. Status: {response.status}")
                return False
    except urllib.error.HTTPError as e:
        print(f"❌ HTTP Error importing '{workflow_name}': {e.code} - {e.reason}")
        try:
            print(e.read().decode('utf-8'))
        except:
            pass
        return False
    except urllib.error.URLError as e:
        print(f"❌ Connection Error: {e.reason}")
        return False
    except Exception as e:
        print(f"❌ Unexpected Error: {e}")
        return False

def main():
    parser = argparse.ArgumentParser(description='Bulk import n8n workflows.')
    parser.add_argument('--dir', default='agents', help='Directory to search for workflows (default: agents)')
    parser.add_argument('--host', help='n8n Host URL (e.g., https://n8n.example.com)')
    parser.add_argument('--key', help='n8n API Key')
    parser.add_argument('--dry-run', action='store_true', help='Scan files without importing')

    args = parser.parse_args()

    # If not dry run, require host and key
    if not args.dry_run and (not args.host or not args.key):
        print("Error: --host and --key are required unless using --dry-run")
        parser.print_help()
        sys.exit(1)

    # Use current working directory + provided dir
    search_dir = os.path.join(os.getcwd(), args.dir)
    if not os.path.exists(search_dir):
        print(f"Error: Directory '{search_dir}' does not exist.")
        sys.exit(1)

    print(f"🔍 Searching for workflows in: {search_dir}")
    files = find_workflow_files(search_dir)
    print(f"Found {len(files)} potential workflow files.\n")

    if not args.dry_run:
        existing_names = get_existing_workflows(args.host, args.key)
        print(f"ℹ️  Found {len(existing_names)} existing workflows on server.")
    else:
        existing_names = set()

    success_count = 0
    skipped_count = 0
    for file_path in files:
        # Check name before processing
        try:
             with open(file_path, 'r', encoding='utf-8') as f:
                wd = json.load(f)
                w_name = wd.get('name', os.path.basename(file_path))
                if w_name in existing_names:
                    print(f"⏭️  Skipping '{w_name}' (already exists)")
                    skipped_count += 1
                    continue
        except:
            pass 

        if import_workflow(file_path, args.host, args.key, args.dry_run):
            success_count += 1
            
    if args.dry_run:
        print(f"\n✨ Dry run complete. Found and validated {success_count} workflows.")
    else:
        print(f"\n✨ Import complete. Imported: {success_count}, Skipped: {skipped_count}, Total Files: {len(files)}")

if __name__ == '__main__':
    main()

File: scripts/import_workflows.js
const fs = require('fs');
const path = require('path');
const https = require('https');

// Configuration
const N8N_BASE_URL = process.env.N8N_BASE_URL || 'https://n8n-production-e313.up.railway.app';
const N8N_API_KEY = process.env.N8N_API_KEY;
const AGENTS_DIR = path.join(__dirname, '../agents');

if (!N8N_API_KEY) {
    console.error('❌ Error: N8N_API_KEY environment variable is required.');
    console.log('Usage: N8N_API_KEY=your_key node scripts/import_workflows.js');
    process.exit(1);
}

// Utility to recursively find JSON files
function findWorkflowFiles(dir, fileList = []) {
    const files = fs.readdirSync(dir);
    
    files.forEach(file => {
        const filePath = path.join(dir, file);
        const stat = fs.statSync(filePath);
        
        if (stat.isDirectory()) {
            findWorkflowFiles(filePath, fileList);
        } else if (file.endsWith('.json')) {
            // Basic validation check - read file content
            try {
                const content = JSON.parse(fs.readFileSync(filePath, 'utf8'));
                if (content.nodes && Array.isArray(content.nodes)) {
                    fileList.push(filePath);
                }
            } catch (e) {
                // Ignore non-JSON or invalid files
            }
        }
    });
    return fileList;
}

// Check if fetch is available (Node 18+), else polyfill with https
const fetchRequest = async (url, method, body = null) => {
    // If global fetch exists, use it
    if (typeof fetch !== 'undefined') {
        const headers = {
            'X-N8N-API-KEY': N8N_API_KEY,
            'Content-Type': 'application/json'
        };
        const options = {
            method,
            headers
        };
        if (body) options.body = JSON.stringify(body);
        
        const response = await fetch(url, options);
        if (!response.ok) {
            const text = await response.text();
            throw new Error(`API Error ${response.status}: ${text}`);
        }
        return response.json();
    } else {
        // Fallback for older Node versions (Manual HTTPS)
        return new Promise((resolve, reject) => {
            const urlObj = new URL(url);
            const options = {
                hostname: urlObj.hostname,
                port: urlObj.port || 443,
                path: urlObj.pathname + urlObj.search,
                method: method,
                headers: {
                    'X-N8N-API-KEY': N8N_API_KEY,
                    'Content-Type': 'application/json'
                }
            };

            const req = https.request(options, (res) => {
                let data = '';
                res.on('data', chunk => data += chunk);
                res.on('end', () => {
                    if (res.statusCode >= 200 && res.statusCode < 300) {
                        try {
                            resolve(JSON.parse(data));
                        } catch (e) {
                            resolve({}); // Handle empty response
                        }
                    } else {
                        reject(new Error(`API Error ${res.statusCode}: ${data}`));
                    }
                });
            });

            req.on('error', reject);
            if (body) req.write(JSON.stringify(body));
            req.end();
        });
    }
};

async function main() {
    console.log(`🚀 Starting bulk import to ${N8N_BASE_URL}...`);
    
    try {
        // 1. Get existing workflows to map names to IDs
        console.log('🔍 Fetching existing workflows...');
        const existingWorkflowsRaw = await fetchRequest(`${N8N_BASE_URL}/api/v1/workflows`, 'GET');
        const existingWorkflows = existingWorkflowsRaw.data || [];
        
        const workflowMap = new Map();
        existingWorkflows.forEach(w => workflowMap.set(w.name, w.id));
        console.log(`✅ Found ${existingWorkflows.length} existing workflows.`);

        // 2. Find local files
        const localFiles = findWorkflowFiles(AGENTS_DIR);
        console.log(`📂 Found ${localFiles.length} workflow files in ${AGENTS_DIR}`);

        // 3. Import each file
        for (const filePath of localFiles) {
            const workflowData = JSON.parse(fs.readFileSync(filePath, 'utf8'));
            const name = workflowData.name || path.basename(filePath, '.json');
            
            // Validate: remove invalid static 'id' fields from nodes if they conflict? 
            // n8n import usually handles this, but API doesn't like workflow ID in body if POSTing
            // We usually post the whole body.
            
            if (workflowMap.has(name)) {
                // UPDATE
                const id = workflowMap.get(name);
                console.log(`🔄 Updating: ${name} (${id})`);
                await fetchRequest(`${N8N_BASE_URL}/api/v1/workflows/${id}`, 'PUT', workflowData);
            } else {
                // CREATE
                console.log(`✨ Creating: ${name}`);
                await fetchRequest(`${N8N_BASE_URL}/api/v1/workflows`, 'POST', workflowData);
            }
        }
        
        console.log('🎉 Bulk import complete!');
        
    } catch (error) {
        console.error('❌ Import failed:', error.message);
        process.exit(1);
    }
}

main();
